pkg/compat/client.go

```diff
@@ -0,0 +1,151 @@
+package compat
+
+import (
+	"context"
+	appv1 "k8s.io/api/apps/v1"
+	appv1beta1 "k8s.io/api/apps/v1beta1"
+	extv1beta1 "k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	dapi "k8s.io/client-go/discovery"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"strconv"
+	"strings"
+)
+
+//
+// A smart client.
+// Provides seamless API version compatibility.
+type Client struct {
+	k8sclient.Client
+	dapi.DiscoveryInterface
+	// major k8s version.
+	Major int
+	// minor k8s version.
+	Minor int
+}
+
+//
+// Create a new client.
+func NewClient(restCfg *rest.Config) (k8sclient.Client, error) {
+	rClient, err := k8sclient.New(
+		restCfg,
+		k8sclient.Options{
+			Scheme: scheme.Scheme,
+		})
+	if err != nil {
+		return nil, err
+	}
+	dClient, err := dapi.NewDiscoveryClientForConfig(restCfg)
+	if err != nil {
+		return nil, err
+	}
+	version, err := dClient.ServerVersion()
+	if err != nil {
+		return nil, err
+	}
+	major, err := strconv.Atoi(version.Major)
+	if err != nil {
+		return nil, err
+	}
+	minor, err := strconv.Atoi(strings.Trim(version.Minor, "+"))
+	if err != nil {
+		return nil, err
+	}
+	nClient := &Client{
+		Client:             rClient,
+		DiscoveryInterface: dClient,
+		Major:              major,
+		Minor:              minor,
+	}
+
+	return nClient, nil
+}
+
+//
+// Down convert a resource as needed based on cluster version.
+func (c Client) downConvert(obj runtime.Object) runtime.Object {
+	if c.Minor < 16 {
+		if _, cast := obj.(*appv1.Deployment); cast {
+			return &appv1beta1.Deployment{}
+		}
+		if _, cast := obj.(*appv1.DeploymentList); cast {
+			return &appv1beta1.DeploymentList{}
+		}
+		if _, cast := obj.(*appv1.DaemonSet); cast {
+			return &extv1beta1.DaemonSet{}
+		}
+		if _, cast := obj.(*appv1.DaemonSetList); cast {
+			return &extv1beta1.DaemonSetList{}
+		}
+	}
+
+	return obj
+}
+
+//
+// Get the specified resource.
+// The resource will be converted to a compatible version as needed.
+func (c Client) Get(ctx context.Context, key k8sclient.ObjectKey, in runtime.Object) error {
+	obj := c.downConvert(in)
+	err := c.Client.Get(ctx, key, obj)
+	if err != nil {
+		return err
+	}
+	if in != obj {
+		err := scheme.Scheme.Convert(obj, in, ctx)
```

Comment:
@jortel you said that this reference to `scheme` has the necessary context to do the conversion correct? I'm surprised we were able to skip passing scheme through, but I'll take your word for it if you've observed this working.

Reply:
The `Manager.GetScheme()` returns the global `scheme.Scheme` when not constructed with a `Scheme` in the _options_.  The manager code generated by kube-builder does not construct with a Scheme.